internal/fs/type_cache_test.go

```diff
@@ -0,0 +1,421 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A collection of tests for a file system where we do not attempt to write to
+// the file system at all. Rather we set up contents in a GCS bucket out of
+// band, wait for them to be available, and then read them via the file system.
+
+package fs_test
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/config"
+	gcsfusefs "github.com/googlecloudplatform/gcsfuse/internal/fs"
+	"github.com/googlecloudplatform/gcsfuse/internal/mount"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/storageutil"
+	"github.com/googlecloudplatform/gcsfuse/internal/util"
+
+	"github.com/jacobsa/oglematchers"
+	. "github.com/jacobsa/ogletest"
+)
+
+// The following is the control of an os.Stat(name) call in case of GCSFuse,
+// for understanding how the tests work. The stat call
+// 1. comes as fs.LookUpInode() call to gcsfuse, which
+// 2. queries type-cache of the parent directory, without the suffix '/'.
+// 2.1 If entry is found in type-cache, then that is returned as type.
+// 2.2 If entry is not found in type-cache, then its type is queried from GCS, and the returned type is stored in type-cache and is also returned as cache.
+// 3. If the input to os.Stat() had a suffix '/' and the return type is not ExplicitDir or ImplicitDir, then an error containing 'not a directory' is returned.
```

Comment:
Can we test this behavior as part of our test? Since, this is the base of our test, we should add a test, where if we add file and directory with the same name, the second stat (either file or directory should throw error).

Reply:
Added 2 tests for this.